mlflow/transformers.py

```diff
@@ -0,0 +1,919 @@
+_SUPPORTED_RETURN_TYPES = {"pipeline", "components"}
+_logger = logging.getLogger(__name__)
+
+
+def _model_packages(model) -> List[str]:
+    """
+    Determines which pip libraries should be included based on the base model engine
+    type.
+
+    :param model: The model instance to be saved in order to provide the required underlying
+                  deep learning execution framework dependency requirements.
+    :return: A list of strings representing the underlying engine-specific dependencies
+    """
+    engine = _get_engine_type(model)
+    if engine == "torch":
+        return ["torch", "torchvision"]
+    else:
+        return [engine]
+
+
+@experimental
+def get_default_pip_requirements(model) -> List[str]:
+    """
+    :param model: The model instance to be saved in order to provide the required underlying
+                  deep learning execution framework dependency requirements. Note that this must
+                  be the actual model instance and not a Pipeline.
+    :return: A list of default pip requirements for MLflow Models that have been produced with the
+             ``transformers`` flavor. Calls to :py:func:`save_model()` and :py:func:`log_model()`
+             produce a pip environment that contain these requirements at a minimum.
+    """
+
+    from transformers import TFPreTrainedModel, FlaxPreTrainedModel, PreTrainedModel
+
+    if not isinstance(model, (TFPreTrainedModel, FlaxPreTrainedModel, PreTrainedModel)):
+        raise MlflowException(
+            "The supplied model type is unsupported. The model must be one of: "
+            "PreTrainedModel, TFPreTrainedModel, or FlaxPreTrainedModel",
+            error_code=INVALID_PARAMETER_VALUE,
+        )
+    base_reqs = ["transformers"]
+    try:
+        base_reqs.extend(_model_packages(model))
+        return [_get_pinned_requirement(module) for module in base_reqs]
+    except RuntimeError:
+        dependencies = [
+            _get_pinned_requirement(module)
+            for module in ["transformers", "torch", "torchvision", "tensorflow"]
+        ]
+        _logger.warning(
+            "Could not infer model execution engine type due to huggingface_hub not "
+            "being installed or unable to connect in online mode. Adding full "
+            f"dependency chain: {dependencies}"
+        )
+        return dependencies
+
+
+def _isinstance_named_tuple(obj):
+    obj_type = type(obj)
+    bases = obj_type.__bases__
+    if len(bases) != 1 or bases[0] != tuple:
+        return False
+    fields = getattr(obj_type, "_fields", None)
+    if not isinstance(fields, tuple):
+        return False
+    return all(type(name) == str for name in fields)
+
+
+def _convert_component_dict_model(model: dict):
+    """
+    Convert a submitted component-based model in a dictionary to a namedtuple
+    """
+    ComponentModel = namedtuple("ComponentModel", model, rename=True)
```

Comment:
I would pull this out from this function so that we can use `isinstance(..., ComponentModel)`

Reply:
After looking at that type validator and how it's not really possible to do the dynamic renaming in a clean way, I deleted these functions and implemented a subclass of NamedTuple. It's a bit more complex, but it gives us the ability to do explicit type validation to ensure that only properly qualified objects are able to be processed in the module.